kong/clustering/wrpc_data_plane.lua

```diff
@@ -0,0 +1,379 @@
+        ngx_log(ngx_ERR, _log_prefix, "unable to pre-create cached config file: ",
+                ffi.string(ffi.C.strerror(ffi.errno())))
+
+      else
+        ffi.C.close(fd)
+      end
+    end
+
+    assert(ngx.timer.at(0, function(premature)
+      self:communicate(premature)
+    end))
+  end
+end
+
+
+local wrpc_config_service
+local function get_config_service()
+  if not wrpc_config_service then
+    wrpc_config_service = wrpc.new_service()
+    wrpc_config_service:add("kong.services.config.v1.config")
+    wrpc_config_service:set_handler("ConfigService.SyncConfig", function(peer, data)
+      if peer.config_semaphore then
+        if data.config.plugins then
+          for _, plugin in ipairs(data.config.plugins) do
+            plugin.config = protobuf.pbunwrap_struct(plugin.config)
+          end
+        end
+        data.config._format_version = data.config.format_version
+        data.config.format_version = nil
+
+        peer.config_obj.next_config = data.config
+        peer.config_obj.next_config_version = tonumber(data.version)
+        if peer.config_semaphore:count() <= 0 then
+          -- the following line always executes immediately after the `if` check
+          -- because `:count` will never yield, end result is that the semaphore
+          -- count is guaranteed to not exceed 1
+          peer.config_semaphore:post()
+        end
+      end
+      return { accepted = true }
+    end)
+  end
+
+  return wrpc_config_service
+end
+
+function _M:communicate(premature)
+
+  if premature then
+    -- worker wants to exit
+    return
+  end
+
+  local conf = self.conf
+
+  -- TODO: pick one random CP
+  local address = conf.cluster_control_plane
+  local log_suffix = " [" .. address .. "]"
+
+  local c = assert(ws_client:new(WS_OPTS))
+  local uri = "wss://" .. address .. "/v1/outlet?node_id=" ..
+              kong.node.get_id() ..
+              "&node_hostname=" .. kong.node.get_hostname() ..
+              "&node_version=" .. KONG_VERSION
+
+  local opts = {
+    ssl_verify = true,
+    client_cert = self.cert,
+    client_priv_key = self.cert_key,
+    protocols = "wrpc.konghq.com",
+  }
+  if conf.cluster_mtls == "shared" then
+    opts.server_name = "kong_clustering"
+  else
+    -- server_name will be set to the host if it is not explicitly defined here
+    if conf.cluster_server_name ~= "" then
+      opts.server_name = conf.cluster_server_name
+    end
+  end
+
+  local reconnection_delay = math.random(5, 10)
+  do
+    local res, err = c:connect(uri, opts)
+    if not res then
+      ngx_log(ngx_ERR, _log_prefix, "connection to control plane ", uri, " broken: ", err,
+        " (retrying after ", reconnection_delay, " seconds)", log_suffix)
+
+      assert(ngx.timer.at(reconnection_delay, function(premature)
+        self:communicate(premature)
+      end))
+      return
+    end
+  end
+
+  local config_semaphore = semaphore.new(0)
+  local peer = wrpc.new_peer(c, get_config_service(), { channel = true })
```

Comment:
What's the best way for me to share a wrpc connection among Ngnix workers? Should each worker create its own peer or should I share the peer among workers?

Reply:
I need to let each Nginx worker push its local data to the cp.

Reply:
one worker owns the connection and creates a peer.  the other workers create a remote_peer, which use a channel to communicate with the "real" one.

right now the channel name is a hardcoded constant, so there can only be one "main" peer (the one created in wrpc_data_plane) and all remote_peers go to that single one.  so anywhere in the code you can use `wrpc.new_remote_peer()` and it will use it.   if we require separate wrpc connections, we'll have to pass the channel name somewhere.

note, the "main" peer doesn't encode/decode the remote_peer's data, so you can add new service definitions and don't have to add to the "main" peer, just the remote_peer

Reply:
I will create another wrpc connection for the telemetry endpoint.

Reply:
There will only be one wrpc connection for each data plane seems to be a very restrictive assumption.

Reply:
then i'll have to add a "connection name" and move the current constant from `wrcp.lua` to `wrpc_data_plane.lua`

Reply:
I am wondering why do you need a channel? It seems that you never create a `remote_client`. no usage of `wprc.new_remote_client`

Reply:
the config handling is done completely in a single worker so it doesn't need it; but the wrpc has to be usable from any worker, so i added channel.lua and new_remote_client()

Reply:
Cool. I was just curious why you enabled the `channel` for the config connection, while it is not needed. https://github.com/Kong/kong/blob/7e9a9a28e82a7a58a3c501f05f241b137f676f69/kong/clustering/wrpc_data_plane.lua#L257

Reply:
it's expected that any code would have to initiate calls to the CP.  this allows any part of code to just do it without having to deal with connection handling.

Reply:
done